### PR TITLE
Importing `metadata` for `Log`-entries properly and importing existing Logs and Comments properly

### LINKED ISSUE #2448: Import of Log entries should be checked for existence based on their UUID
We have to see what happens when a node is re-imported with more log entries - We should import the new log entries based on their UUID.

This is a continuation of the work done for issue #2423 under PR #2447

At least a test should be added that demonstrated that if the same node is imported with more log entries, only the non existing log entries should be added to the new DB.

### LINKED ISSUE #2484: Imported Log entries cannot be re-exported due to metadata column content
When importing `Log` entries (tested with a django backend), the dict in the `metadata` column in the `db_dblog` table is imported with single-quotations instead of double-quotations, making it impossible to re-export the `Log`s again, if one would want to do that.

When `Log`s are originally created, the dict in `metadata` uses double-quotations.

### LINKED ISSUE #2489: Importing "old" Comments for existing Node
Related to #2448.

Similar tests as those related to #2448 and PR #2486.

Make sure `Comment`s related to existing `Node` are correctly imported.

Make sure that if one changes a previously exported `Comment` in their DB, the imported (old) `Comment` with the same UUID, should not overwrite the existing one.
This should not happen for `Log`s in anyway, since they should/cannot be changed once made (@sphuber this is true, right?).

### PR DESCRIPTION
**Pull request by CasperWA**
#### Fixes #2448 
**Short recap**: Check re-import of `Log`s is done properly.
**Wanted feature**: Import of `Log`s must not duplicate existing `Log`s, based on UUID.
**Example**: A node with 2 `Log`s exist in the database. I export the node and its `Log`s. Later, a new `Log` message is added to the same node in the database. I import my "old" export of the node (containing the 2 original `Log`s). I *must* end up with the node and its 3 `Log`s in my database, i.e. there should be no change in the content of the database concerning the existing node and its 3 `Log` entries.

Tests added:

- `test_export_and_import.py#TestLogs#`**`test_multiple_imports_for_single_node`**
- `test_export_and_import.py#TestLogs#`**`test_reimport_of_logs_for_single_node`**

#### Fixes #2484
**Short recap**: Fix import of `metadata` of `Log`s with django backend.

Changes made:

- Make a type-check of `metadata` in `orm/logs.py#__init__`
- Make sure a JSON-serializable string is produced for `metadata` when importing with a django backend.

Tests added or updated:

- `orm/test_logs.py#`**`test_raise_wrong_metadata_type_error`**
- `test_export_and_import.py#TestLogs#`**`test_export_import_of_critical_log_msg_and_metadata`** (updated)
- `test_export_and_import.py#TestLogs#`**`test_export_of_imported_logs`**

#### Fixes #2489
**Short recap**: Similar to issue #2448 and #2486, but for `Comment`s.
**Wanted feature**: Import of `Comment`s must not duplicate existing `Comment`s, based on UUID.
However, if a `Comment` has been edited, it should be possible to choose whether to keep/import the *newest* `Comment`, based on modification time, or *overwrite* existing `Comment`s with those being imported.

Changes made:

- When importing, a new kwarg can be passed: `comment_mode`. This is checked for what to do, when a `Comment` with the same UUID as an existing `Comment` is in the import-file. The following values are recognized:
  
  - `newest`: Keep/import the newest `Comment`, based on the modification time, `mtime`.
  - `overwrite`: Import the `Comment`, overwriting the existing `Comment`.
  
  In actuality, and for efficiency purposes, when importing, the existing `Comment` will always be updated with the values of the imported `Comment`, instead of having the existing `Comment` deleted before creating a new `Comment` in the database.

  - `orm/importexport.py#`**`_merge_comment`** handles the logic for `comment_mode`.

- `verdi import` understands `comment_mode`.
- Make sure a `str` is returned for the UUID when using the django backend.
- Several changes done by @szoupanos to do with making sure `mtime` is not changed upon ex-/import.
  These include:

  - a contextmanager: `backends/djsite/db/models.py#`**`suppress_auto_now`**,
  - an `__init__` for `backends/sqlaclhemy/models/comment.py`,
  - an `mtime` check for `Node`s in `backends/sqlalchemy/models/node.py`,
  - specifying `ctime` and `mtime` in `__init__` of `orm/implementation/django/comments.py` as well as in `../nodes.py` and similarly for the SQLAlchemy models.

Tests added:

- `test_export_and_import.py#TestComments#`**`test_multiple_imports_for_single_node`**
- `test_export_and_import.py#TestComments#`**`test_reimport_of_comments_for_single_node`**
- `test_export_and_import.py#TestComments#`**`test_mtime_of_imported_comments`**
- `test_export_and_import.py#TestComments#`**`test_import_arg_comment_mode`**
- `orm/implementation/`**`test_comments.py`**
- `orm/implementation/test_nodes#`**`test_creation_with_time`**

#### Minor fixes
Several minor fixes were done, especially in `importexport.py`.

### CODE REVIEW

**Comment by CasperWA** [comment]
The issue #2448 is not fixed for SQLAlchemy backend.

**Comment by CasperWA** [comment]
@sphuber and @szoupanos, I have finished getting Logs and Comments to import correctly with existing entities present in the db.

However, the code is probably quite dirty, and especially in the two import functions.
There is a point, where I believe a comparison of `mtime` for Comments will never be true. There is a comment in the code after the comparison is made.
The reason I think so, is due to the nature of `mtime` that will be updated upon import, if I'm not wrong? So the "original" `mtime` is lost. Instead one may compare `ctime`, but they should in theory always be the same anyway.

File: aiida/orm/implementation/sqlalchemy/logs.py [hunk 1]
```diff
@@ -113,9 +113,11 @@ def delete_many(self, filters):
         """
         Delete all log entries in the table
         """
+        session = get_scoped_session()
+
         if not filters:
             for entry in models.DbLog.query.all():
                 entry.delete()
-            get_scoped_session().commit()
+            session.commit()
```
Comment (szoupanos) [inline aiida/orm/implementation/sqlalchemy/logs.py]: I would put his in a transaction
https://docs.sqlalchemy.org/en/latest/orm/session_transaction.html

```
Session = sessionmaker(bind=engine, autocommit=True)
session = Session()
session.begin()
try:
    item1 = session.query(Item).get(1)
    item2 = session.query(Item).get(2)
    item1.foo = 'bar'
    item2.bar = 'foo'
    session.commit()
except:
    session.rollback()
    raise
```

File: aiida/orm/implementation/django/comments.py [hunk 2]
```diff
@@ -116,3 +116,13 @@ def delete(self, comment):
...
         except ObjectDoesNotExist:
             raise exceptions.NotExistent("Comment with id '{}' not found".format(comment))
+
+    # pylint: disable=no-self-use
+    def delete_many(self, filters):
+        """
+        Delete all comment entries in the table
+        """
+        if not filters:
+            models.DbComment.objects.all().delete()
```
Comment (szoupanos) [inline aiida/orm/implementation/django/comments.py]: Maybe this should also be in a transaction

File: aiida/orm/importexport.py [hunk 3]
```diff
@@ -729,7 +761,57 @@ def import_data_dj(in_path, user_group=None, ignore_unknown_nodes=False,
                 for import_entry_id, entry_data in existing_entries[model_name].items():
                     unique_id = entry_data[unique_identifier]
                     existing_entry_id = foreign_ids_reverse_mappings[model_name][unique_id]
+                    import_data = dict(deserialize_field(
+                        k, v, fields_info=fields_info,
+                        import_unique_ids_mappings=import_unique_ids_mappings,
+                        foreign_ids_reverse_mappings=foreign_ids_reverse_mappings)
+                                       for k, v in entry_data.items())
                     # TODO COMPARE, AND COMPARE ATTRIBUTES
+
```
Comment (szoupanos) [inline aiida/orm/importexport.py]: This code has very few things that are backend dependent - it can go easily to a function that can be shared by both backend import functions

**Comment by CasperWA** [comment]
As a note:
At the current state there is an issue concerning `mtime` and the django backend. The correct import of identical Comments (samme UUID, but perhaps different content due to an edit) relies on comparing the modification time (`mtime`) and determining which is newest. The "old" Comment will then be dismissed.
This works with SQLA.
For django, it is another story. As it is now, `mtime` is flushed upon initializing a Comment. However, undoing this flush, I could not find a way to update `mtime` upon editing a Comment's content.
In short, this means that `mtime` changes upon import with a django backend, and therefore cannot be relied upon as a measure of which Comment is the "newest".

If one knows how to set up the django `DateTimeField` for `mtime` in such a way that it is equivalent to how SQLA has defined `mtime`, this would solve the issue. Otherwise, I will have to write a separate import logic for Comments when using a django backend. Principally, this is something I do not wish to do.

**Comment by szoupanos** [comment]
I looked at the ctime and mtime of a node and its corresponding comment and they are correctly exported in both backends. So the problem of mtime update in Django is somewhere else.

I will keep looking.

Django
```
(aiidapy2) aiida@ubuntu_vm1_new:~/spyros_test/dj$ verdi -p issue_1759_dj shell
-/home/aiida/aiidapy2/local/lib/python2.7/site-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)
Python 2.7.15rc1 (default, Nov 12 2018, 14:31:15)
Type "copyright", "credits" or "license" for more information.

IPython 5.8.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: from aiida.orm.nodes.data.data import Data
   ...: d = Data()
   ...: d.store()
   ...: d.add_comment('lalala')
   ...: d.store()
   ...: c = d.get_comments()[0]
   ...: print(d.ctime)
   ...: print(d.mtime)
   ...: print(c.ctime)
   ...: print(c.mtime)
   ...:
2019-02-26 18:23:09.463472+00:00
2019-02-26 18:23:09.468064+00:00
2019-02-26 18:23:09.491641+00:00
2019-02-26 18:23:09.492302+00:00


(aiidapy2) aiida@ubuntu_vm1_new:~/spyros_test/dj$ verdi -p issue_1759_dj export create lala_dj.zip -N 1
-/home/aiida/aiidapy2/local/lib/python2.7/site-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)
STARTING EXPORT...
STORING DATABASE ENTRIES...
Exporting a total of 3 db entries, of which 1 nodes.
STORING NODE ATTRIBUTES...
STORING NODE EXTRAS...
STORING NODE LINKS...
STORING GROUP ELEMENTS...
STORING DATA...
STORING FILES...
File written in      0.191 s.
Success: wrote the export archive file to lala_dj.zip
(aiidapy2) aiida@ubuntu_vm1_new:~/spyros_test/dj$ unzip lala_dj.zip
Archive:  lala_dj.zip
  inflating: data.json
  inflating: metadata.json
   creating: nodes/e6/8c/99ef-cec4-45b8-aa92-91c8795ece88/path/
(aiidapy2) aiida@ubuntu_vm1_new:~/spyros_test/dj$ cat data.json | pjson
{
    "links_uuid": [

    ],
    "export_data": {
        "Node": {
            "1": {
                "ctime": "2019-02-26T18:23:09.463472",
                "description": "",
                "dbcomputer": null,
                "process_type": null,
                "label": "",
                "user": 1,
                "mtime": "2019-02-26T18:23:09.468064",
                "nodeversion": 1,
                "type": "data.Data.",
                "public": false,
                "uuid": "e68c99ef-cec4-45b8-aa92-91c8795ece88"
            }
        },
        "Comment": {
            "1": {
                "dbnode": 1,
                "ctime": "2019-02-26T18:23:09.491641",
                "content": "lalala",
                "user": 1,
                "mtime": "2019-02-26T18:23:09.492302",
                "uuid": "a36daf57-b80e-4677-a57d-1d30c0ccbe3a"
            }
        },
        "User": {
            "1": {
                "first_name": "AiiDA",
                "last_name": "Daemon",
                "email": "aiida@localhost",
                "institution": ""
            }
        }
    },
    "node_attributes": {
        "1": {

        }
    },
    "node_extras": {
        "1": {
            "_aiida_hash": "42758129f68829d4c15bfd29fcfed830b051b14d4e62646aff063b5e910aab89"
        }
    },
    "groups_uuid": {

    },
    "node_extras_conversion": {
        "1": {
            "_aiida_hash": null
        }
    },
    "node_attributes_conversion": {
        "1": {

        }
    }
}
```

SQLA

```
(aiidapy2) aiida@ubuntu_vm1_new:~/spyros_test/sqla$ verdi -p issue_1759_sqla shell
-/home/aiida/aiidapy2/local/lib/python2.7/site-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)
Python 2.7.15rc1 (default, Nov 12 2018, 14:31:15)
Type "copyright", "credits" or "license" for more information.

IPython 5.8.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: from aiida.orm.nodes.data.data import Data
   ...: d = Data()
   ...: d.store()
   ...: d.add_comment('lalala')
   ...: d.store()
   ...: c = d.get_comments()[0]
   ...: print(d.ctime)
   ...: print(d.mtime)
   ...: print(c.ctime)
   ...: print(c.mtime)
   ...:
2019-02-26 18:23:53.755727+00:00
2019-02-26 18:23:53.777706+00:00
2019-02-26 18:23:53.780637+00:00
2019-02-26 18:23:53.780654+00:00

(aiidapy2) aiida@ubuntu_vm1_new:~/spyros_test/sqla$ verdi -p issue_1759_sqla export create lala_sqla.zip -N 1
-/home/aiida/aiidapy2/local/lib/python2.7/site-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)
STARTING EXPORT...
STORING DATABASE ENTRIES...
Exporting a total of 3 db entries, of which 1 nodes.
STORING NODE ATTRIBUTES...
STORING NODE EXTRAS...
STORING NODE LINKS...
STORING GROUP ELEMENTS...
STORING DATA...
STORING FILES...
File written in      0.191 s.
Success: wrote the export archive file to lala_sqla.zip
(aiidapy2) aiida@ubuntu_vm1_new:~/spyros_test/sqla$ unzip lala_sqla.zip
Archive:  lala_sqla.zip
  inflating: data.json
  inflating: metadata.json
   creating: nodes/ec/6a/97d1-e607-4489-a11e-d9a6307c7e3d/path/
(aiidapy2) aiida@ubuntu_vm1_new:~/spyros_test/sqla$ cat data.json | pjson
{
    "links_uuid": [

    ],
    "export_data": {
        "Node": {
            "1": {
                "ctime": "2019-02-26T18:23:53.755727",
                "description": "",
                "dbcomputer": null,
                "process_type": null,
                "label": "",
                "user": 1,
                "mtime": "2019-02-26T18:23:53.777706",
                "nodeversion": 1,
                "type": "data.Data.",
                "public": false,
                "uuid": "ec6a97d1-e607-4489-a11e-d9a6307c7e3d"
            }
        },
        "Comment": {
            "1": {
                "dbnode": 1,
                "ctime": "2019-02-26T18:23:53.780637",
                "content": "lalala",
                "user": 1,
                "mtime": "2019-02-26T18:23:53.780654",
                "uuid": "8696ee17-a7c8-4029-b73a-57f9363ad924"
            }
        },
        "User": {
            "1": {
                "first_name": "AiiDA",
                "last_name": "Daemon",
                "email": "aiida@localhost",
                "institution": ""
            }
        }
    },
    "node_attributes": {
        "1": {

        }
    },
    "node_extras": {
        "1": {
            "_aiida_hash": "42758129f68829d4c15bfd29fcfed830b051b14d4e62646aff063b5e910aab89"
        }
    },
    "groups_uuid": {

    },
    "node_extras_conversion": {
        "1": {
            "_aiida_hash": null
        }
    },
    "node_attributes_conversion": {
        "1": {

        }
    }
}
```

**Comment by szoupanos** [comment]
I solved the "problem" in Django with a context manager that disable the mtime update, for all the imported models, during the creation of objects in import.

I am looking why the mtime in SQLA is not updated during import for Comments (which is the behaviour that we want) but it is updated for Nodes... (!!!)

**Comment by szoupanos** [comment]
And the onupdate of SQLA has the behaviour that we would like
https://docs.sqlalchemy.org/en/latest/core/metadata.html#sqlalchemy.schema.Column.params.onupdate

To be seen what happens with Nodes

**Comment by CasperWA** [comment]
> I solved the "problem" in Django with a context manager that disable the mtime update, for all the imported models, during the creation of objects in import.

This is great news! Then this PR should be ready to go.

> I am looking why the mtime in SQLA is not updated during import for Comments (which is the behaviour that we want) but it is updated for Nodes... (!!!)

Ah, then perhaps this should be fixed in another PR?

For future reference (and this will/should be added to the documentation as well). The behavior we want is the following:
`mtime` should _only_ update when the columns of a database table is changed. For a Comment this means for example an edit/change of the content updates `mtime`.
At no time should either export or import change `mtime`. Not for Comment, not for Node. Not for anything.

**Comment by szoupanos** [comment]
@CasperWA 
I also fixed the SQLA Node issue & I also extended the corresponding BackendEntities (Node & Comment) to correctly support mtime/ctime conventions in a homogeneous manner.
The code needs a bit of cleaning (not only mine). But everything should be OK.
Let's see if you can take over this tomorrow.

@sphuber 
Why the BackendComment has a setter method for the comment and not a property setter as BackendNode has for label?

File: aiida/orm/implementation/sqlalchemy/logs.py (hunk 1 above)
Comment (CasperWA) [reply aiida/orm/implementation/sqlalchemy/logs.py]: I will do the following:

```python
session = get_scoped_session()

try:
    # do the if, else statement
except Exception as exc:
    session.rollback()
    raise exceptions.DbContentError(exc)
```

File: aiida/orm/implementation/django/comments.py (hunk 2 above)
Comment (CasperWA) [reply aiida/orm/implementation/django/comments.py]: How do I do this with Django?

File: aiida/orm/importexport.py (hunk 3 above)
Comment (CasperWA) [reply aiida/orm/importexport.py]: I partly agree - although there are some potential db entry deletions.
Should I try to create a common function or leave it as it?

**Comment by CasperWA** [comment]
Sorry @szoupanos. It seems your named contributions have disappeared doing the rebasing. This was not my intention. However, your changes/additions are still in what I have force-pushed. I don't understand why the rebase doesn't acknowledge you as well though, this used to be the case when I have previously rebased.

**Comment by szoupanos** [comment]
> Sorry @szoupanos. It seems your named contributions have disappeared doing the rebasing. This was not my intention. However, your changes/additions are still in what I have force-pushed. I don't understand why the rebase doesn't acknowledge you as well though, this used to be the case when I have previously rebased.

If you feel very uncomfortable, there is a solution :D

Creating a commit with multiple authors
https://help.github.com/en/articles/creating-a-commit-with-multiple-authors

File: aiida/orm/implementation/sqlalchemy/comments.py [hunk 4]
```diff
@@ -109,11 +125,24 @@ def delete(self, comment):
...
+
+    # pylint: disable=no-self-use
+    def delete_many(self, filters):
+        """
+        Delete all comment entries in the table
+        """
+        session = get_scoped_session()
+
+        if not filters:
+            for entry in models.DbComment.query.all():
```
Comment (szoupanos) [inline aiida/orm/implementation/sqlalchemy/comments.py]: This will be very likely inefficient. Maybe you can combine the deletion with the query
https://docs.sqlalchemy.org/en/latest/orm/query.html#sqlalchemy.orm.query.Query.delete
Just check that we don't violate anything (written in the warnings - I think not from a first look)

File: aiida/orm/implementation/django/comments.py (hunk 2 above)
Comment (szoupanos) [reply aiida/orm/implementation/django/comments.py]: There is an example at the aiida.orm.importexport.import_data_dj
But maybe a transaction is not needed here. It is an one line statement/command and I hope that this is executed as one deletion sql statement (there is not reason not to). In such a case I think that the transaction will not help - sorry for the initial comment

File: aiida/orm/implementation/sqlalchemy/comments.py (hunk 4 above)
Comment (odarbelaeze) [reply aiida/orm/implementation/sqlalchemy/comments.py]: Deleting the entries one by one is the conventional way to go in sqlalchemy, deleting in bulk has caveats like not running cascade effects in python and possibly contaminating the state of the session. On the other hand, bulk delete is actually more performant, but I'd say those performance gains become important only when deleting a huge amount of objects.

Comment (CasperWA) [reply aiida/orm/implementation/sqlalchemy/comments.py]: Reading the docs referenced by @szoupanos I understand that the cascading is still observed as long as it is defined, when the column containing the ForeignKey is defined. This should be the true for all relevant cases, meaning this should be fine.
On another note, I can see that I forgot to do a try, escape for Comments (only did it for Logs).

Comment (CasperWA) [reply aiida/orm/implementation/sqlalchemy/comments.py]: Additionally, for Comments the associated Node or User should never be deleted (however a Comment may be deleted, when deleting a Node).
I will delete in bulk then, using `session.query(models.DbComment).delete()` and similarly for Logs.

File: aiida/orm/implementation/sqlalchemy/logs.py (hunk 1 above)
Comment (szoupanos) [reply aiida/orm/implementation/sqlalchemy/logs.py]: Something that I missed when I did my previous comment is that a commit in this method (at least in SQLA), will commit everything until this point. This is something that we don't want during import because it will commit data in the middle of the import.

A solution is nested transactions. Martin has implemented a context manager for transactions that seems to correctly handle nested transactions for both backends.
aiida.orm.implementation.backends.Backend#transaction

This could be used which will also allow you to have less code overlap in these two delete_many method.

aiida.orm.implementation.backends.Backend#transaction could also be used (please test) to replace the current transactions in import_data_dj/import_data_sqla

File: aiida/orm/importexport.py [hunk 5]
```diff
@@ -405,58 +405,59 @@ def merge_extras(old_extras, new_extras, mode):
...
-    
-    elif mode == 'knl': # 'none': keep old extras, do not add imported ones
+
+    if mode == 'knl': # 'none': keep old extras, do not add imported ones
         return old_extras
-    
-    elif mode == 'kcu': # 'update_existing' operation: if an extra already exists,
-                               # overwrite its new value with a new one
+
+    if mode == 'kcu': # 'update_existing' operation: if an extra already exists,
```
[the remark this answers is not in the document]
Comment (CasperWA) [reply aiida/orm/importexport.py]: because it returns in the previous two if's

File: aiida/backends/tests/orm/test_logs.py [hunk 6]
```diff
@@ -248,3 +247,71 @@ def test_log_querybuilder(self):
...
+            'name': 'aiida.orm.node.process.calculation.CalculationNode'
+        }
+
+        # str of dict metadata
+        wrong_metadata_format = str(correct_metadata_format)
+
+        # JSON-serialized-deserialized dict metadata
+        json_metadata_format = json.loads(json.dumps(correct_metadata_format))
+
+        # Check no error is raised when creating a Log
```
[the remark this answers is not in the document]
Comment (CasperWA) [reply aiida/backends/tests/orm/test_logs.py]: Yup. comment changed

File: aiida/orm/logs.py [hunk 7]
```diff
@@ -105,6 +105,9 @@ def __init__(self, time, loggername, levelname, dbnode_id, message='', metadata=
         """Construct a new log"""
         from aiida.common import exceptions
 
+        if metadata is not None and not isinstance(metadata, dict):
+            raise TypeError("metadata must be a dict")
```
[the remark this answers is not in the document]
Comment (CasperWA) [reply aiida/orm/logs.py]: A check for a Log's metadata to be equal before and after import will be done.

**Comment by CasperWA** [comment]
After discussing with @sphuber we realized the `_merge_comment` function should not actually delete entries in the existing db, but instead simply update the entries with the new values...

Therefore, I have also moved all changes to `delete_many` methods to a new branch and out of this PR.

Furthermore, "experiments" using the backend.transaction method should wait to be implemented until after the beta. So this will also be moved to a new branch and out of this PR.

File: aiida/orm/importexport.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (szoupanos) [reply aiida/orm/importexport.py]: I would use if at the first comparison and elif/else in the subsequent comparisons (since, it doesn't make the code more complex and you operate on the same variable)
But this is something minor and I suppose personal preference, so leave it as it is

**Comment by szoupanos** [comment]
> After discussing with @sphuber we realized the `_merge_comment` function should not actually delete entries in the existing db, but instead simply update the entries with the new values...
> 
> Therefore, I have also moved all changes to `delete_many` methods to a new branch and out of this PR.
> 
> Furthermore, "experiments" using the backend.transaction method should wait to be implemented until after the beta. So this will also be moved to a new branch and out of this PR.

I agree, if you update the existing values, there is no need to delete them and re-insert them.
I would keep this work though and merge it to 1b or 1 because  I suppose that it would be useful somewhere else.
